File: src/scripts/salary_slip.js
// Copyright (c) 2019, Libermatic and contributors
// For license information, please see license.txt

export const salary_slip = {
  set_training_data: async function(frm) {
    await frappe.call({
      method:
        'psd_customization.fitness_world.api.salary_slip.set_trainings_in_salary_slip',
      args: { doc_json: frm.doc, set_in_response: 1 },
    });
  },
  calculate_training_earnings: async function(frm) {
    const {
      salary_structure,
      total_training_months = 0,
      training_rate = 0,
    } = frm.doc;
    const {
      message: { training_salary_component } = {},
    } = await frappe.db.get_value(
      'Salary Structure',
      salary_structure,
      'training_salary_component'
    );
    const earning = frm.doc.earnings.find(
      ({ salary_component }) => salary_component === training_salary_component
    );
    if (earning) {
      earning.amount = total_training_months * training_rate;
      frm.refresh_field('earnings');
    }
  },
};

function calculate_training_months(frm) {
  frm.set_value(
    'total_training_months',
    frm.doc.trainings.map(({ months }) => months).reduce((a, x = 0) => a + x, 0)
  );
}

export const salary_slip_training = {
  training: async function(frm, cdt, cdn) {
    const { training } = frappe.get_doc(cdt, cdn);
    if (training) {
      const no_of_occurences = frm.doc.trainings.filter(
        ({ training: t }) => t === training
      ).length;
      if (no_of_occurences > 1) {
        frm.get_field('trainings').grid.grid_rows_by_docname[cdn].remove();
      } else {
        const { start_date, end_date } = frm.doc;
        const {
          message: { months, gym_subscription } = {},
        } = await frappe.call({
          method:
            'psd_customization.fitness_world.api.salary_slip.get_training_data',
          args: { training, start_date, end_date },
        });
        frappe.model.set_value(cdt, cdn, 'months', months);
        frappe.model.set_value(cdt, cdn, 'subscription', gym_subscription);
      }
    } else {
      ['months', 'subscription'].forEach(field => {
        frappe.model.set_value(cdt, cdn, field, null);
      });
    }
  },
  months: function(frm, cdt, cdn) {
    calculate_training_months(frm);
  },
  trainings_remove: function(frm, cdt, cdn) {
    calculate_training_months(frm);
  },
};

export default { salary_slip, salary_slip_training };

File: psd_customization/fitness_world/api/salary_slip.py
# -*- coding: utf-8 -*-
# Copyright (c) 2019, Libermatic and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.utils import getdate, cint
from builtins import str
import json
from toolz import merge, pluck


@frappe.whitelist()
def set_trainings_in_salary_slip(doc_json, set_in_response=0):
    doc = frappe.get_doc(json.loads(doc_json)) \
        if isinstance(doc_json, str) else doc_json
    if not doc.salary_structure:
        joining_date, relieving_date = frappe.db.get_value(
            'Employee', doc.employee, ['date_of_joining', 'relieving_date']
        )
        doc.salary_structure = doc.check_sal_struct(
            joining_date, relieving_date
        )
    structure = frappe.get_doc('Salary Structure', doc.salary_structure)
    doc.salary_slip_based_on_training = \
        structure.salary_slip_based_on_training
    doc.set('trainings', [])
    if doc.salary_slip_based_on_training:
        trainings = get_trainings_for_salary_slip(
            doc.employee, doc.end_date
        )
        for row in trainings:
            doc.append('trainings', {
                'training': row.name,
                'member_name': row.gym_member_name,
                'months': row.months,
                'subscription': row.gym_subscription,
            })
        doc.total_training_months = sum(pluck('months', trainings))
        doc.training_rate = structure.training_monthly_rate
        add_earning_for_training(
            doc,
            structure.training_salary_component,
            doc.total_training_months * doc.training_rate
        )
    if cint(set_in_response):
        frappe.response.docs.append(doc)


def get_trainings_for_salary_slip(employee, end_date):
    trainer = frappe.db.exists('Gym Trainer', {
        'employee': employee
    })
    if not trainer:
        return []
    trainings = frappe.db.sql(
        """
            SELECT
                ta.name AS name,
                s.member_name AS gym_member_name,
                s.name AS gym_subscription,
                ta.salary_till AS salary_till,
                ta.from_date AS from_date,
                ta.to_date AS to_date,
                s.day_fraction AS day_fraction
            FROM `tabTrainer Allocation` AS ta
            LEFT JOIN `tabGym Subscription` AS s
                ON s.name = ta.gym_subscription
            WHERE
                ta.gym_trainer = %(trainer)s AND
                IFNULL(
                    ta.salary_till,
                    DATE_SUB(ta.from_date, INTERVAL 1 DAY)
                ) < %(end_date)s
        """,
        values={
            'trainer': trainer,
            'end_date': end_date,
        },
        as_dict=1,
    )
    return map(_set_days(end_date), trainings)


def add_earning_for_training(doc, salary_component, amount):
    for row in doc.earnings or []:
        if row.salary_component == salary_component:
            row.amount = amount
            return
    doc.append('earnings', {
        'salary_component': salary_component,
        'abbr': frappe.db.get_value(
            'Salary Component', salary_component, 'salary_component_abbr'
        ),
        'amount': amount,
    })


def _set_days(end_date):
    def fn(row):
        from_date = row.salary_till + 1 if row.salary_till else row.from_date
        to_date = min(getdate(end_date), row.to_date)
        days = (to_date - from_date).days + 1
        return frappe._dict(
            merge(row, {
                'months': days * row.day_fraction,
            })
        )
    return fn